src/superblock.rs

```diff
@@ -0,0 +1,262 @@
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use crate::checksum::Checksum;
+use crate::error::{Corrupt, Ext4Error, Incompatible};
+use crate::features::{IncompatibleFeatures, ReadOnlyCompatibleFeatures};
+use crate::util::{read_u16, read_u32, u64_from_hilo};
+
+/// Information about the filesystem.
+#[derive(Clone, Debug, Eq, PartialEq)]
+#[repr(C)]
+pub struct Superblock {
+    pub(crate) block_size: u32,
+    pub(crate) blocks_count: u64,
+    pub(crate) inode_size: u16,
+    pub(crate) inodes_per_block_group: u32,
+    pub(crate) block_group_descriptor_size: u16,
+    pub(crate) num_block_groups: u32,
+    pub(crate) incompatible_features: IncompatibleFeatures,
+    pub(crate) read_only_compatible_features: ReadOnlyCompatibleFeatures,
+    pub(crate) checksum_seed: u32,
+}
+
+impl Superblock {
+    /// Size (in bytes) of the superblock on disk.
+    pub(crate) const SIZE_IN_BYTES_ON_DISK: usize = 1024;
+
+    /// Construct `Superblock` from bytes.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the length of `bytes` is less than
+    /// [`Self::SIZE_IN_BYTES_ON_DISK`].
+    pub(crate) fn from_bytes(bytes: &[u8]) -> Result<Self, Ext4Error> {
+        assert!(bytes.len() >= Self::SIZE_IN_BYTES_ON_DISK);
+
+        // OK to unwrap: already checked the length.
+        let s_blocks_count_lo = read_u32(bytes, 0x4);
+        let s_first_data_block = read_u32(bytes, 0x14);
+        let s_log_block_size = read_u32(bytes, 0x18);
+        let s_blocks_per_group = read_u32(bytes, 0x20);
+        let s_inodes_per_group = read_u32(bytes, 0x28);
+        let s_magic = read_u16(bytes, 0x38);
+        let s_inode_size = read_u16(bytes, 0x58);
+        let s_feature_incompat = read_u32(bytes, 0x60);
+        let s_feature_ro_compat = read_u32(bytes, 0x64);
+        let s_uuid = &bytes[0x68..0x68 + 16];
+        let s_desc_size = read_u16(bytes, 0xfe);
+        let s_blocks_count_hi = read_u32(bytes, 0x150);
+        let s_checksum_seed = read_u32(bytes, 0x270);
+        let s_checksum = read_u32(bytes, 0x3fc);
+
+        let blocks_count = u64_from_hilo(s_blocks_count_hi, s_blocks_count_lo);
+
+        let block_size = 2u32.pow(10 + s_log_block_size);
+
+        if s_magic != 0xef53 {
+            return Err(Ext4Error::Corrupt(Corrupt::SuperblockMagic));
+        }
+
+        let incompatible_features = check_incompat_features(s_feature_incompat)
+            .map_err(Ext4Error::Incompatible)?;
+        let read_only_compatible_features =
+            ReadOnlyCompatibleFeatures::from_bits_retain(s_feature_ro_compat);
+
+        let num_block_groups = u32::try_from(
+            (blocks_count - u64::from(s_first_data_block)
+                + u64::from(s_blocks_per_group)
+                - 1)
+                / u64::from(s_blocks_per_group),
```

Comment:
The `(n + d - 1) / d` trick is used to bias the division up in case its not an even division. But, turns out this isn't needed in Rust: https://doc.rust-lang.org/std/primitive.u64.html#method.div_ceil

Switched the code to use `div_ceil` and added a comment.